pkg/util/watch.go

```diff
@@ -493,6 +428,82 @@ func Watch[C WatchClient[L], L metav1.ListMetaAccessor, T any, P WatchObject[T]]
 	return &store, nil
 }
 
+// helper for Watch. Error events are expected to already have been handled by the caller.
+func handleEvent[T any, P ~*T](
+	store *WatchStore[T],
+	config WatchConfig,
+	handlers WatchHandlerFuncs[P],
+	eventType watch.EventType,
+	meta metav1.Object,
+	ptr P,
+) error {
+	uid := meta.GetUID()
+	name := NamespacedName{Namespace: meta.GetNamespace(), Name: meta.GetName()}
+	obj := (*T)(ptr)
+
+	// Some of the cases below don't actually require locking the store. Most of the events that we
+	// recieve *do* though, so we're better off doing it here for simplicity.
```

Comment:
could have each case in the switch statement be: 
```go
return func() error {
// lock handling
// current code
}
```
I think it'd be worth it, given how contended this lock may be.

Reply:
Realistically we'll be fine. The only "normal" case that won't require locking is if we receive a "bookmark" event, which shouldn't be very often. (the other cases are things that we're not expecting, so will panic)